pkg/operator/resource/resourcemerge/admissionregistration.go

```diff
@@ -0,0 +1,76 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// ExpectedMutatingWebhooksConfiguration returns last applied generation for MutatingWebhookConfiguration resource registered in operator
+func ExpectedMutatingWebhooksConfiguration(name string, previousGenerations []operatorsv1.GenerationStatus) int64 {
+	generation := GenerationFor(previousGenerations, schema.GroupResource{Group: admissionregistrationv1.SchemeGroupVersion.Group, Resource: "mutatingwebhookconfigurations"}, "", name)
+	if generation != nil {
+		return generation.LastGeneration
+	}
+	return -1
+}
+
+// SetMutatingWebhooksConfigurationGeneration updates operator generation status list with last applied generation for provided MutatingWebhookConfiguration resource
+func SetMutatingWebhooksConfigurationGeneration(generations *[]operatorsv1.GenerationStatus, actual *admissionregistrationv1.MutatingWebhookConfiguration) {
+	if actual == nil {
+		return
+	}
+	SetGeneration(generations, operatorsv1.GenerationStatus{
+		Group:          admissionregistrationv1.SchemeGroupVersion.Group,
+		Resource:       "mutatingwebhookconfigurations",
+		Name:           actual.Name,
+		LastGeneration: actual.ObjectMeta.Generation,
+	})
+}
+
+// ExpectedValidatingWebhooksConfiguration returns last applied generation for ValidatingWebhookConfiguration resource registered in operator
+func ExpectedValidatingWebhooksConfiguration(name string, previousGenerations []operatorsv1.GenerationStatus) int64 {
+	generation := GenerationFor(previousGenerations, schema.GroupResource{Group: admissionregistrationv1.SchemeGroupVersion.Group, Resource: "validatingwebhookconfigurations"}, "", name)
+	if generation != nil {
+		return generation.LastGeneration
+	}
+	return -1
+}
+
+// SetValidatingWebhooksConfigurationGeneration updates operator generation status list with last applied generation for provided ValidatingWebhookConfiguration resource
+func SetValidatingWebhooksConfigurationGeneration(generations *[]operatorsv1.GenerationStatus, actual *admissionregistrationv1.ValidatingWebhookConfiguration) {
+	if actual == nil {
+		return
+	}
+	SetGeneration(generations, operatorsv1.GenerationStatus{
+		Group:          admissionregistrationv1.SchemeGroupVersion.Group,
+		Resource:       "validatingwebhookconfigurations",
+		Name:           actual.Name,
+		LastGeneration: actual.ObjectMeta.Generation,
+	})
+}
+
+// EnsureValidatingWebhooksConfiguration ensures that the existing matches the required.
+// modified is set to true when existing had to be updated with required.
+func EnsureValidatingWebhooksConfiguration(modified *bool, existing *admissionregistrationv1.ValidatingWebhookConfiguration, required admissionregistrationv1.ValidatingWebhookConfiguration) {
+	EnsureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+
+	// we stomp everything
+	if !equality.Semantic.DeepEqual(existing.Webhooks, required.Webhooks) {
+		*modified = true
+		existing.Webhooks = required.Webhooks
+	}
```

Comment:
Is there any defaulting within the Webhooks that we need to account for here? We could end up in a hot loop if there's server side defaulting that we are missing

Reply:
Having to call defaulting in here is the reason for possible hot loops. If we pull server implementation for webhooks, someone would have to suffer a burden of maintaining it in sync with upstream, or otherwise, any defaulting implementation change would be a 100% cause for hot looping. The location @sttts pointed at, is an internal `kubenetes/kubernetes` package. If it is not exposed at https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/api/admissionregistration/v1, I don't think it is wise to use it and especially, copying that. I'm not sure that the strict rules for kubernetes api changes are applied there, so the probability of described situation is higher. Let's leave it to the server-side apply, as a future improvement.

Right now I could agree that if you immediately apply the updated resource, returned from `ApplyMutatingWebhookConfiguration` or `ApplyValidatingWebhookConfiguration`once again, it would not be filtered by the logic, and `Update` call will still happen. Even if the resource is technically the same, from the algorithmic perspective you provided a different input, and the change should occur.

I'm adding the missing comment from `resourceapply/apps.go` from where the implementation was inspired, so the explanation will be preserved: https://github.com/openshift/library-go/blob/3b4f06f8efeda380cd78d17b09ed6d72edc1b06a/pkg/operator/resource/resourceapply/apps.go#L21-L51

Reply:
While I agree with that comment for the deployment implementation, this does not match with the logic that you have in this implementation.

We are checking the `Webhooks` for equality here, we aren't solely relying on the spec hash. There is no way that the two sets of webhooks will be equal unless the input has already been defaulted, so this will currently cause a hot loop. 

We have two options, we copy the logic from the K8s package (definitely do not import it), or we remove the lines I've highlighted here and solely rely on the spec hash annotation for detecting differences.

> someone would have to suffer a burden of maintaining it in sync with upstream, or otherwise, any defaulting implementation change would be a 100% cause for hot looping.

This has already been discussed, this is what Stefan was saying about in his previous comments. We have to be careful when adding to this package because if we don't get the defaulting right, we will hot loop. He's already said it's annoying because we have to keep them up to date. See https://github.com/openshift/library-go/pull/836#issuecomment-659425034

Reply:
Now I see it, thanks. Updated.